File: Reference MATLAB Files/EnduranceTest.m
clear all
close all

%% Settings
name = 'Debugging';

freq = 2100;
TestDuration = 10; % Duration of the test
MeasFreq = 1; % time interval between measurements (approximative)

Lrec = 10000; % Recording length
Fs = 1e6; % Sampling freq

ampIn = 12; % input amplitude in V
ampRange = 2; % oscillo amplitude in V


%% Getting the data
system('taskkill /im MultiChannel.exe'); % Close TiePie in case it was open

if verLessThan('matlab', '8')
    error('Matlab 8.0 (R2012b) or higher is required.');
end

% Open LibTiePie and display library info if not yet opened:
import LibTiePie.Const.*
import LibTiePie.Enum.*

if ~exist('LibTiePie', 'var')
    % Open LibTiePie:
    LibTiePie = LibTiePie.Library;
end

% Enable network search:       
LibTiePie.Network.AutoDetectEnabled = true;

% Search for devices:
LibTiePie.DeviceList.update();

% figure('Name','Live plot')

figure('Name','Live plot')
vamprms = [];
ave = []; % average signal value
DatStor = {};



item = LibTiePie.DeviceList.getItemByIndex(0);
gen = item.openGenerator();


    
% Generator settings:

        % Set signal type:
        gen.SignalType = ST.SINE;

        % Set frequency:
        gen.Frequency = freq; 

        % Set amplitude:
        gen.Amplitude = ampIn; % 2 V

        % Set offset:
        gen.Offset = 0; % 0 V

        % Enable output:
        gen.OutputOn = true;
        
        % recover some information from generator
        % vGenFreq(incre) = gen.Frequency;

        % Start signal generation:
        gen.start();


tic
duration = 0;
vDuration = [];
while duration < TestDuration

%     for k = 0 : LibTiePie.DeviceList.Count - 1
%         item = LibTiePie.DeviceList.getItemByIndex(k);
%         if item.canOpen(DEVICETYPE.OSCILLOSCOPE) && item.canOpen(DEVICETYPE.GENERATOR)
%             scp = item.openOscilloscope();
%             if ismember(MM.BLOCK, scp.MeasureModes)
%                 gen = item.openGenerator();
%                 break;
%             else
%                 clear scp;
%             end
%         end
%     end
    item = LibTiePie.DeviceList.getItemByIndex(0);
    scp = item.openOscilloscope();
    
    clear item
    
    if exist('scp', 'var') && exist('gen', 'var')
        % Oscilloscope settings:

        % Set measure mode:
        scp.MeasureMode = MM.BLOCK;

        % Set sample frequency:
        scp.SampleFrequency = Fs; % 1 MHz

        % Set record length:
        scp.RecordLength = Lrec; % 10000 Samples

        % Set pre sample ratio:
        scp.PreSampleRatio = 0; % 0 %

        % For all channels:
        for ch = scp.Channels
            % Enable channel to measure it:
            ch.Enabled = true;

            % Set range:
            ch.Range = ampRange; 

            % Set coupling:
            ch.Coupling = CK.DCV; % DC Volt

            % Release reference:
            clear ch;
        end

        % Set trigger timeout:
        scp.TriggerTimeOut = 1; % 1 s

        % Disable all channel trigger sources:
        for ch = scp.Channels
            ch.Trigger.Enabled = false;
            clear ch;
        end

        % Locate trigger input:
        triggerInput = scp.getTriggerInputById(TIID.GENERATOR_NEW_PERIOD); % or TIID.GENERATOR_START or TIID.GENERATOR_STOP

        if triggerInput == false
            clear triggerInput;
            clear scp;
            clear gen;
            error('Unknown trigger input!');
        end

        % Enable trigger input:
        triggerInput.Enabled = true;

        % Release reference to trigger input:
        clear triggerInput;

        % Start measurement:
        scp.start();

        % Wait for measurement to complete:
        while ~scp.IsDataReady
            pause(10e-3) % 10 ms delay, to save CPU time.
        end

        % Get data:
        arData = scp.getData();
        ave = [ave; mean(arData)];
        arData = arData - mean(arData);

        % Get all channel data value ranges (which are compensated for probe gain/offset):
        clear darRangeMin;
        clear darRangeMax;
        for i = 1 : length(scp.Channels)
            [darRangeMin(i), darRangeMax(i)] = scp.Channels(i).getDataValueRange();
        end

        % Plot results:
%         if mod(incre,40) == 1
%             figure
%             plot((1:scp.RecordLength) / scp.SampleFrequency, arData);
%             xlabel('Time [s]');
%             ylabel('Amplitude [V]');
%             legend('Channel 1','Channel 2');
%             grid on;
%         end

        % Close oscilloscope:
        clear scp;

    else
        error('No oscilloscope available with block measurement support or generator available in the same unit!');
    end

    DatStor = {DatStor arData};
    
    % Plot
    amprms = rms(arData-mean(arData));
    vamprms = [vamprms; amprms];
    
    duration = toc;
    vDuration = [vDuration duration];
    plot(vDuration, vamprms*1e3)
    grid on;
    xlabel('Duration [s]');
    ylabel('Oscillation amplitude [mV]');
    legend('Channel 1','Channel 2');
    
%     plot(Vvolt,rms(arData-mean(arData)));
%     grid on;
%     xlabel('Input piezo voltage (to be converted in length) [V]');
%     ylabel('Output voltage [mV]');
%     legend('Channel 1','Channel 2')

    pause(MeasFreq);
end

% Stop generator:
gen.stop();

% Disable output:
gen.OutputOn = false;

% Close generator:
clear gen;


save([name,' - Endurance Test - ',datestr(now, 'dd-mmm-yyyy HH-MM-SS')]);

clear all % necessary to use TiePie after MatLab

system('cd C:\Users\cri\Documents\Etienne & start MatLab.tps'); % Reopen TiePie

return

%% Post process 

% % Amplitude
% vamp = [];
% for k = 1:length(DatStor)
%     amp = max(DatStor{1,k})-min(DatStor{1,k});
%     vamp = [vamp; amp];
% end
% 
% % Plot
% figure('Name','min max')
% plot(vfreq,vamp,'-o')
% grid on;
% xlabel('Frequency [Hz]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');

% FFT
vfour = [];
for k = 1:length(DatStor)
    four = abs(fft(DatStor{1,k}-mean(DatStor{1,k})));
    four = four(1:Lrec/2+1,:);
    vfour = [vfour four];
end

% Plot
% figure('Name','fft')
% plot(Fs*(0:Lrec/2)/Lrec*1e-3,vfour);
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude');


% RMS
vamprms = [];
for k = 1:length(DatStor)
    amprms = rms(DatStor{1,k}-mean(DatStor{1,k}));
    vamprms = [vamprms; amprms];
end

% Plot
figure('Name','RMS')
plot(vfreq/1e3,vamprms*1e3)
grid on;
xlabel('Frequency [kHz]');
ylabel('Oscillation amplitude [mV]');
legend('Channel 1','Channel 2');

% Plot
figure('Name','Average signal')
plot(vfreq/1e3,ave*1e3)
grid on;
xlabel('Frequency [kHz]');
ylabel('Amplitude [mV]');
legend('Channel 1','Channel 2');

mean(ave)

% % Plot
normal = vamprms(:,1)./vamprms(:,2);
% figure('Name','RMS normalised')
% plot(vfreq/1e3,normal)
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplification factor');

%% Peaks
figure('Name','RMS peaks')
findpeaks(normal,vfreq/1e3,'MinPeakHeight',2*mean(normal))
grid on;
xlabel('Frequency [kHz]');
ylabel('Amplification factor');

[pks,loc] = findpeaks(normal,vfreq/1e3,'MinPeakHeight',2*mean(normal));
cond = sum(loc >= (target-tolerance)/1e3 & loc <= (target+tolerance)/1e3);
if cond == 0
    disp('Failed chip')
else
    disp('Passed chip')
end


%% test
vamp = [];
for k = 1:length(DatStor)
    amp = mean(DatStor{1,k});
    vamp = [vamp; amp];
end

% % Plot
% figure('Name','amp mean')
% plot(vfreq/1e3,vamp-mean(vamp))
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');
% 
% % Plot
% figure('Name','amp mean 2')
% plot(vfreq/1e3,vamp)
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');

% % RMS
% vamprms = [];
% for k = 1:length(DatStor)
%     amprms = rms(DatStor{1,k}(Lrec/2:end,:)-mean(DatStor{1,k}(Lrec/2:end,:)));
%     vamprms = [vamprms; amprms];
% end

% % Plot
% figure('Name','RMS')
% plot(vfreq/1e3,vamprms)
% grid on;
% xlabel('Frequency [kHz]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');
% 
% % FFT
% vfour = [];
% for k = 1:length(DatStor)
%     four = abs(fft(DatStor{1,k}(Lrec/2:end,:)-mean(DatStor{1,k}(Lrec/2:end,:))));
%     four = four(1:Lrec/4+1,:);
%     vfour = [vfour four];
% end
% 
% % Plot
% figure('Name','fft')
% plot(Fs*(0:Lrec/4)/Lrec*2,vfour);
% grid on;
% xlabel('Frequency [Hz]');
% ylabel('Amplitude');

save([name,' - ',datestr(now, 'dd-mmm-yyyy HH-MM-SS')]);
loc

%% Finder 
% [f,dummy] = ginput(1);
% point = round((f*1e3 - start)*(points-1)/(stop-start) + 1);
% 
% % Plot
% figure
% plot((0:Lrec-1)/Fs, DatStor{1,point});
% xlabel('Time [s]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');
% grid on;

%% Test 07/05/2021
% pos = 90;
% 
% figure
% plot((1:Lrec) / Fs, DatStor{1,pos}-mean(DatStor{1,pos}));
% xlabel('Time [s]');
% ylabel('Amplitude [V]');
% legend('Channel 1','Channel 2');
% grid on;

File: python-libtiepie-master/Python Routines/EnduranceTest.py
# -*- coding: utf-8 -*-
"""
Created on Wednesday 10.11.2021 at 09:35

Miraex - Endurance Test for piezoelectric shaker

@author: Claudio Jaramillo
"""

from __future__ import print_function
import sys
import os
from typing import final
import libtiepie
import time
import datetime

import numpy as np
import MiraexLib.plot as miraex_plt
import MiraexLib.misc as miraex_misc
from MiraexLib.printinfo import*

# %% Magic numbers and setup
_seconds = 1
_minutes = _seconds*60
_hours = _minutes*60
_days = _hours*24

'''
myDays = input('Test duration : days ? ')
myHours = input('Test duration : hours ? ')
myMinutes = input('Test duration : minutes ? ')
mySeconds = input('Test duration : seconds ? ')
'''

nb_of_graphs_to_show = int(input(
    'For how many epochs would you like to show the dynamic plot ? '))

# %% Settings
freq = 2e3

TestLoopDuration = 10*_seconds  # Duration of the test loop
TestTotalDuration = 0*_days + 1*_minutes  # Duration of the entire test

if TestTotalDuration <= TestLoopDuration:
    print('ERROR')
    print('The total test duration MUST be larger than the loop duration.')
    sys.exit(1)

MeasFreq = 1  # time interval between measurements (approximative)

Lrec = 100  # Recording length : the number of samples that will be taken per loop, default = 10000
Fs = 200e3  # Sampling freq

ampOut = 1  # input amplitude in V
ampRange = 4  # oscilloscope amplitude in V

# %% Setting up result file
# Create file_name
name = 'EnduranceTest'
file_name = miraex_misc.CreateFileName(name, 'txt')
writeDir_func = miraex_misc.CreateWriteDir('raw')

# Open the text file to write the results
result_file = open(writeDir_func+file_name, "ab")
print(result_file)

result_file.truncate(0)

# %% Opening devices and setting up devices

# Print library info:
print_library_info()

# Enable network search:
libtiepie.network.auto_detect_enabled = True

# Search for devices:
libtiepie.device_list.update()

# Try to open a generator:
gen = None
for item in libtiepie.device_list:
    if item.can_open(libtiepie.DEVICETYPE_GENERATOR):
        gen = item.open_generator()
        if gen:
            break

'''
REMOVE COMMENTS FOR HS3
'''

if gen:
    try:
        # Set signal type:
        gen.signal_type = libtiepie.ST_SINE

        # Set frequency:
        gen.frequency = freq  # in Hz

        # Set amplitude:
        gen.amplitude = ampOut  # in V

        # Set offset:
        gen.offset = 0  # in V

        # Enable output:
        gen.output_on = True

        # Print generator info:
        print_device_info(gen)

        # Start signal generation:
        gen.start()

    except:
        print('Exception at Signal Generation')

# %% Perform test

# Setting starting counters
total_duration = 0
epoch_counter = 0

# Setting starting time
total_start_time = time.time()
print('Starting time : ', total_start_time)

# outer loop
while total_duration < TestTotalDuration:

    # Computing the number of epochs
    epoch_counter += 1
    print('Begin of epoch ', epoch_counter)

    # creation of empty lists and variables that will be modified in the inner loop and need to be reset at each outer loop
    loop_counter = 0
    loop_duration = 0
    loop_start_time = time.time()

    average_list = []
    data_no_offset = []
    Vect_AmpRMS = []
    Vect_Duration = []

    print(f'{loop_duration = }')
    print(f'{TestLoopDuration = }')
    # Inner loop
    while loop_duration < TestLoopDuration:

        # Counting the number of loops
        print('--------------------------')
        loop_counter += 1
        print('Begin of inner loop : ', loop_counter)

        # Try to open an oscilloscope with block measurement support:
        scp = None
        for item in libtiepie.device_list:
            if item.can_open(libtiepie.DEVICETYPE_OSCILLOSCOPE):
                scp = item.open_oscilloscope()
                if scp.measure_modes & libtiepie.MM_BLOCK:
                    break
                else:
                    scp = None

        if scp:
            try:
                # Set measure mode:
                scp.measure_mode = libtiepie.MM_BLOCK

                # Set sample frequency:
                scp.sample_frequency = Fs  # in Hz

                # Set record length:
                scp.record_length = Lrec  # nb of samples

                # Set pre sample ratio:
                scp.pre_sample_ratio = 0  # 0 %

                # For all channels:
                for ch in scp.channels:
                    # Enable channel to measure it:
                    ch.enabled = True

                    # Set range:
                    ch.range = ampRange  # in V

                    # Set coupling:
                    ch.coupling = libtiepie.CK_DCV  # DC Volt

                # Set trigger timeout:
                scp.trigger_time_out = 1  # in secs

                # Disable all channel trigger sources:
                for ch in scp.channels:
                    ch.trigger.enabled = False

                '''
                REMOVE COMMENT FOR HS3
                '''

                # Locate trigger input:
                # or TIID_GENERATOR_START or TIID_GENERATOR_STOP
                trigger_input = scp.trigger_inputs.get_by_id(
                    libtiepie.TIID_GENERATOR_NEW_PERIOD)

                if trigger_input is None:
                    raise Exception('Unknown trigger input!')

                # Enable trigger input:
                trigger_input.enabled = True

                # Print oscilloscope info (optional):
                # print_device_info(scp)

                # Start measurement:
                scp.start()

                # Wait for measurement to complete:
                while not scp.is_data_ready:
                    time.sleep(0.01)  # 10 ms delay, to save CPU time

                # Get data:
                data = scp.get_data()

            except:
                print('Error at Scope Generation')

        # Data processing

        # print(f'{data = }')
        # np_data will be rewritten at every loop iteration
        # np_data[nb of channels of the scope][nb of samples in channel]

        np_data = np.array(data)  # numpy array
        #print(f'{np_data.shape = }')

        # average value of the data fetched during 1 loop per channel
        # The first index is the iteration index
        # The second index is the avg for channel index
        average_list.append([np.mean(np_data[i]) for i in range(len(np_data))])

        # Substract the mean of np_data to each element of np_data during 1 loop
        '''
        Processed data will have n number of elements, each has nb_of_channels sub-dimensions that represent the nb of channels that we are working with in the oscilloscope.

        The number of elements n is the number of iterations of the loop. It increases as the test time increases.
        '''
        # By substracting the mean we remove the DC offset for each channel
        data_no_offset.append(
            [np_data[i] - average_list[loop_counter-1][i] for i in range(len(np_data))])

        del np_data

        '''
        print('data_no_offset')
        print(data_no_offset)
        print('processed data shape = ', np.array(data_no_offset).shape)
        '''

        # Get all channel data value ranges (which are compensated for probe gain/offset)
        data_range_min_max = []
        for channel in scp.channels:
            data_range_min_max.append(
                [channel._get_data_value_min(), channel._get_data_value_max()])

        '''
        COMMENT DEPRECATED, BUT LEAVING IT HERE JUST IN CASE. SAFE TO IGNORE.

        Possible problem here because we are already computing the variance between the datapoints
        and the average of those datapoints. Yet we later take that data anc compute an RMS of it, which
        should yield the STD, but we are computing the STD of the STD ?

        MATLAB reference code lacks documentation, therefore we decide to use the STD of the data as a
        statistic metric and come back to it once the issue has been cleared.
        '''

        # Creation of lists to plot
        '''
        Possible confusion over statistics here in the reference MATLAB code ?

        Here we diverge from the ref. MATLAB file. We will compute the RMS value of the channels 1 and 2 overone acquisition period and write it to a JSON file. We will also plot the RMS values of channels 1 and 2 over that acquisition in order to verify the intergrity of the data and then we simply dump the plot.
        '''
        # RMS of the data fetched in 1 loop iteration, for each channel
        AmpRMS = [np.sqrt(np.mean((np.array(data_no_offset[loop_counter-1][i]))**2))
                  for i in range(len(data_no_offset[loop_counter-1]))]

        # Create a vector with all the RMS values of all loops, updates each loop
        Vect_AmpRMS.append(AmpRMS)
        Vect_AmpRMS_np = np.array(Vect_AmpRMS)

        # Create a Channel first array of RMS values for plotting and better addressing of the data
        Channel_AmpRMS = np.transpose(Vect_AmpRMS_np)

        # Printing data for verification
        '''
        print('Iteration RMS values : ')
        print(Vect_AmpRMS_np)

        print('Channel RMS values : ')
        print(Channel_AmpRMS)
        '''

        # duration before plotting, might be inaccurate to IRL !!!
        loop_duration = time.time() - loop_start_time  # secs

        # creation of duration array for plotting
        Vect_Duration.append(loop_duration)

        # Plotting
        """Dynamic plotting every iteration of the while loop"""
        if nb_of_graphs_to_show >= epoch_counter:
            for i in range(len(Channel_AmpRMS)):
                miraex_plt.DynamicPlot2(Vect_Duration, (Channel_AmpRMS[i]),
                                        'Duration [s]', 'RMS', 'Dynamic Endurance Test')

        else:
            miraex_plt.CloseAllPlots()

        print('End of inner loop ', loop_counter)

    # Writing of RMS values at the end of each EPOCH, the data we write contains all the RMS values for all the loops contained within one epoch

    # f.write(b"\n")
    # In the resulting file each column represents one channel
    np.savetxt(result_file, Vect_AmpRMS_np)

    total_duration = time.time() - total_start_time
    print('END OF EPOCH', epoch_counter)
    print('--------------------------')

'''
REMOVE COMMENT FOR HS3
'''
# Stop generator:
gen.stop()

# Disable output:
gen.output_on = False


# print(VAmpRMS)
print(f'{total_duration = } seconds')

# Don't forget to close the file !
result_file.close()

# %% Post processing
# Required time to write the data !
time.sleep(1)

# Opening data file to plot final data
result_file = open(writeDir_func+file_name, "ab")
final_data = np.loadtxt(writeDir_func+file_name)

# Creating final data
final_data_tp = np.transpose(final_data)

# Plotting final data
x_data = np.arange(final_data_tp.shape[1])

for i in range(len(final_data_tp)):

    legend = 'Channel '+str(i+1)
    miraex_plt.GenericPlot(
        x_data, final_data_tp[i], 'Time', 'RMS', file_name, legend)

# Keep the ShowPlots command at the end of the script !!!!!!
miraex_plt.ShowPlots()
result_file.close()
